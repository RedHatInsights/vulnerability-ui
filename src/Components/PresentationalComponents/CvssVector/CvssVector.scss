.severity-info-box {
    > div {
        h6 {
            margin: 0;
        }
        > div:nth-child(3) {
            h6 {
                .ins-c-skeleton {
                    max-height: 30px !important;
                    max-width: 30% !important;
                }
            }
            div > .ins-c-skeleton {
                max-height: 25px !important;
                max-width: 80% !important;
            }
        }
    }
}
