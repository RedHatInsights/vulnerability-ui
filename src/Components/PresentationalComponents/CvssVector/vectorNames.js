import { createIntl, createIntlCache } from 'react-intl';
import messages from '../../../Messages';

const cache = createIntlCache();
const locale = navigator.language;
const intl = createIntl({
    // eslint-disable-next-line no-console
    onError: console.log,
    locale
}, cache);

export const V3names = {
    AV: {
        name: intl.formatMessage(messages.vectorNamesV3AV),
        values: {
            N: intl.formatMessage(messages.vectorValuesNetwork),
            A: intl.formatMessage(messages.vectorValuesAdjacent),
            L: intl.formatMessage(messages.vectorValuesLocal),
            P: intl.formatMessage(messages.vectorValuesPhysical)
        }
    },
    AC: {
        name: intl.formatMessage(messages.vectorNamesV3AC),
        values: {
            L: intl.formatMessage(messages.vectorValuesLow),
            H: intl.formatMessage(messages.vectorValuesHigh)
        }
    },
    PR: {
        name: intl.formatMessage(messages.vectorNamesV3PR),
        values: {
            N: intl.formatMessage(messages.vectorValuesNone),
            L: intl.formatMessage(messages.vectorValuesLow),
            H: intl.formatMessage(messages.vectorValuesHigh)
        }
    },
    UI: {
        name: intl.formatMessage(messages.vectorNamesV3UI),
        values: {
            N: intl.formatMessage(messages.vectorValuesNone),
            R: intl.formatMessage(messages.vectorValuesRequired)
        }
    },
    S: {
        name: intl.formatMessage(messages.vectorNamesV3S),
        values: {
            U: intl.formatMessage(messages.vectorValuesUnchanged),
            C: intl.formatMessage(messages.vectorValuesChanged)
        }
    },
    C: {
        name: intl.formatMessage(messages.vectorNamesV3C),
        values: {
            N: intl.formatMessage(messages.vectorValuesNone),
            L: intl.formatMessage(messages.vectorValuesLow),
            H: intl.formatMessage(messages.vectorValuesHigh)
        }
    },
    I: {
        name: intl.formatMessage(messages.vectorNamesV3I),
        values: {
            N: intl.formatMessage(messages.vectorValuesNone),
            L: intl.formatMessage(messages.vectorValuesLow),
            H: intl.formatMessage(messages.vectorValuesHigh)
        }
    },
    A: {
        name: intl.formatMessage(messages.vectorNamesV3A),
        values: {
            N: intl.formatMessage(messages.vectorValuesNone),
            L: intl.formatMessage(messages.vectorValuesLow),
            H: intl.formatMessage(messages.vectorValuesHigh)
        }
    },
    CVSS: {
        name: intl.formatMessage(messages.vectorNamesCvssVersion),
        values: {
            '3.0': 3.0
        }
    }
};

export const V2names = {
    AV: {
        name: intl.formatMessage(messages.vectorNamesV2AV),
        values: {
            N: intl.formatMessage(messages.vectorValuesNetwork),
            A: intl.formatMessage(messages.vectorValuesAdjacentNetwork),
            L: intl.formatMessage(messages.vectorValuesLocal)
        }
    },
    AC: {
        name: intl.formatMessage(messages.vectorNamesV2AC),
        values: {
            N: intl.formatMessage(messages.vectorValuesNone),
            L: intl.formatMessage(messages.vectorValuesLow),
            H: intl.formatMessage(messages.vectorValuesHigh)
        }
    },
    Au: {
        name: intl.formatMessage(messages.vectorNamesV2AU),
        values: {
            M: intl.formatMessage(messages.vectorValuesMultiple),
            S: intl.formatMessage(messages.vectorValuesSingle),
            N: intl.formatMessage(messages.vectorValuesNone)
        }
    },
    C: {
        name: intl.formatMessage(messages.vectorNamesV2C),
        values: {
            C: intl.formatMessage(messages.vectorValuesComplete),
            P: intl.formatMessage(messages.vectorValuesPartial),
            N: intl.formatMessage(messages.vectorValuesNone)
        }
    },
    I: {
        name: intl.formatMessage(messages.vectorNamesV2I),
        values: {
            C: intl.formatMessage(messages.vectorValuesComplete),
            P: intl.formatMessage(messages.vectorValuesPartial),
            N: intl.formatMessage(messages.vectorValuesNone)
        }
    },
    A: {
        name: intl.formatMessage(messages.vectorNamesV2A),
        values: {
            C: intl.formatMessage(messages.vectorValuesComplete),
            P: intl.formatMessage(messages.vectorValuesPartial),
            N: intl.formatMessage(messages.vectorValuesNone)
        }
    }
};
