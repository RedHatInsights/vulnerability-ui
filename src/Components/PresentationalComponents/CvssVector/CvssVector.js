import { Popover, Text, TextContent, TextVariants } from '@patternfly/react-core';
import { OutlinedQuestionCircleIcon } from '@patternfly/react-icons';
import { Table, TableBody, TableHeader, TableVariant } from '@patternfly/react-table';
import propTypes from 'prop-types';
import React from 'react';
import { CVEPageContext } from '../../SmartComponents/CVEPage/CVEPage';
import WithLoader from '../WithLoader/WithLoader';
import './CvssVector.scss';
import { injectIntl } from 'react-intl';
import messages from '../../../Messages';
import { V3names, V2names } from './vectorNames';

export const parseVector = (vector, namesMapping) => {
    let res = {};
    vector &&
        vector.split('/').forEach(curr => {
            let [key, val] = curr.split(':');
            const name = namesMapping[key] && namesMapping[key].name;
            const value = (namesMapping[key] && namesMapping[key].values[val]) || 'N/A';
            if (name) {
                res[name] = value;
            }
        });
    return res;
};

const CvssVector = props => {
    const { cvss2_metrics: CVSS2Vector, cvss3_metrics: CVSS3Vector, intl } = props;
    const cvssVer = (CVSS3Vector && 'CVSS 3.0') || (CVSS2Vector && 'CVSS 2.0') || 'CVSS 3.0';
    const cvssVector = CVSS3Vector || CVSS2Vector;
    const namesMapping = (CVSS3Vector && V3names) || (CVSS2Vector && V2names);
    const parsedVector = parseVector(cvssVector, namesMapping);
    const notAvailable = 'N/A';
    // eslint-disable-next-line no-unused-vars
    const { cvssVersion, ...values } = parsedVector;

    return (
        <CVEPageContext.Consumer>
            {context => (
                <TextContent>
                    <WithLoader loading={context.isLoading}>
                        <Popover
                            position="bottom"
                            maxWidth={'100%'}
                            enableFlip
                            headerContent={ `${cvssVer} ${intl.formatMessage(messages.cvssVectorPopoverTitle)}` }
                            bodyContent={
                                cvssVector ?
                                    (
                                        <Table
                                            aria-label={'Metric breakdown'}
                                            variant={TableVariant.compact}
                                            gridBreakPoint={null}
                                            cells={
                                                [intl.formatMessage(messages.cvssVectorMetric),
                                                    intl.formatMessage(messages.cvssVectorValue)]
                                            }
                                            rows={Object.entries(values)}
                                        >
                                            <TableHeader />
                                            <TableBody />
                                        </Table>
                                    ) : (notAvailable)
                            }
                        >
                            <React.Fragment>
                                <Text component={TextVariants.h6} style={{ display: 'inline-block', cursor: 'pointer' }}>
                                    {cvssVer} {intl.formatMessage(messages.cvssVectorVectorString)}
                                    <OutlinedQuestionCircleIcon color={'var(--pf-global--secondary-color--100)'}/>
                                </Text>
                                <br />
                            </React.Fragment>
                        </Popover>
                    </WithLoader>

                    <WithLoader loading={context.isLoading}>{cvssVector || notAvailable} </WithLoader>
                </TextContent>
            )}
        </CVEPageContext.Consumer>
    );
};

CvssVector.propTypes = {
    cvss2_metrics: propTypes.string,
    cvss3_metrics: propTypes.string,
    intl: propTypes.any
};

export default injectIntl(CvssVector);
