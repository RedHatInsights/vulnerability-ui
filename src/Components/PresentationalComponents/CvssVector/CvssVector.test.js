import CvssVector, { parseVector } from './CvssVector';
import { V3names, V2names } from './vectorNames';
import toJson from 'enzyme-to-json';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';

describe('CvssVector', () => {
    it('Should render without parameters', () => {
        const vector = mount(<IntlProvider locale="en"><CvssVector /></IntlProvider>);
        expect(toJson(vector)).toMatchSnapshot();
    });
    it('Should render with CVSSv2', () => {
        const vector = mount(<IntlProvider locale="en"><CvssVector cvss2_metrics="AV:L/AC:H/Au:N/C:N/I:N/A:P" /></IntlProvider>);
        expect(toJson(vector)).toMatchSnapshot();
    });
    it('Should render with CVSSv3', () => {
        const vector = mount(<IntlProvider locale="en"><CvssVector cvss3_metrics="CVSS:3.0/AV:A/AC:L/PR:L/UI:R/S:U/C:L/I:N/A:N" /></IntlProvider>);
        expect(toJson(vector)).toMatchSnapshot();
    });
    it('Should render CVSSv3 when given both', () => {
        const vector = mount(
            <IntlProvider locale="en"><CvssVector cvss2_metrics="AV:L/AC:H/Au:N/C:N/I:N/A:P" cvss3_metrics="CVSS:3.0/AV:A/AC:L/PR:L/UI:R/S:U/C:L/I:N/A:N"/> </IntlProvider>);
        expect(toJson(vector)).toMatchSnapshot();
    });
    it('Should parse CVSSv2', () => {
        const parsed = parseVector('AV:L/AC:H/Au:N/C:N/I:N/A:P', V2names);
        expect(parsed).toMatchObject({
            'Access vector': 'Local',
            'Access complexity': 'High',
            Authentication: 'None',
            'Confidentiality impact': 'None',
            'Integrity impact': 'None',
            'Availability impact': 'Partial'
        });
    });
    it('Should parse CVSSv3', () => {
        const parsed = parseVector('CVSS:3.0/AV:A/AC:L/PR:L/UI:R/S:U/C:L/I:N/A:N', V3names);
        expect(parsed).toMatchObject({
            cvssVersion: 3,
            'Attack vector': 'Adjacent',
            'Attack complexity': 'Low',
            'Privileges required': 'Low',
            'User interaction': 'Required',
            Scope: 'Unchanged',
            Confidentiality: 'Low',
            Integrity: 'None',
            Availability: 'None'
        });
    });
    it('Should parse invalid vector and not fail', () => {
        const parsed = parseVector('CVSS:3.0/AV:A/AC:L/XX:XX', V3names);
        expect(parsed).toMatchObject({ cvssVersion: 3, 'Attack vector': 'Adjacent', 'Attack complexity': 'Low' });
    });
});
