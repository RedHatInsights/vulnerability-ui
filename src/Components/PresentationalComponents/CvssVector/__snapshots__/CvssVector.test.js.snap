// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvssVector Should render CVSSv3 when given both 1`] = `
<injectIntl(CvssVector)
  cvss2_metrics="AV:L/AC:H/Au:N/C:N/I:N/A:P"
  cvss3_metrics="CVSS:3.0/AV:A/AC:L/PR:L/UI:R/S:U/C:L/I:N/A:N"
>
  <CvssVector
    cvss2_metrics="AV:L/AC:H/Au:N/C:N/I:N/A:P"
    cvss3_metrics="CVSS:3.0/AV:A/AC:L/PR:L/UI:R/S:U/C:L/I:N/A:N"
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatDateToParts": [Function],
        "formatDisplayName": [Function],
        "formatHTMLMessage": [Function],
        "formatList": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatNumberToParts": [Function],
        "formatPlural": [Function],
        "formatRelativeTime": [Function],
        "formatTime": [Function],
        "formatTimeToParts": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getDisplayNames": [Function],
          "getListFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralRules": [Function],
          "getRelativeTimeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "default.cancel": "Cancel",
          "default.delete": "Delete",
          "default.remove": "Remove",
          "default.save": "Save",
        },
        "onError": [Function],
        "textComponent": Symbol(react.fragment),
        "timeZone": undefined,
      }
    }
  >
    <TextContent>
      <div
        className="pf-c-content"
      >
        <WithLoader
          loading={true}
        >
          <Skeleton
            isDark={false}
            size="lg"
          >
            <div
              className="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </Skeleton>
        </WithLoader>
        <WithLoader
          loading={true}
        >
          <Skeleton
            isDark={false}
            size="lg"
          >
            <div
              className="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </Skeleton>
        </WithLoader>
      </div>
    </TextContent>
  </CvssVector>
</injectIntl(CvssVector)>
`;

exports[`CvssVector Should render with CVSSv2 1`] = `
<injectIntl(CvssVector)
  cvss2_metrics="AV:L/AC:H/Au:N/C:N/I:N/A:P"
>
  <CvssVector
    cvss2_metrics="AV:L/AC:H/Au:N/C:N/I:N/A:P"
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatDateToParts": [Function],
        "formatDisplayName": [Function],
        "formatHTMLMessage": [Function],
        "formatList": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatNumberToParts": [Function],
        "formatPlural": [Function],
        "formatRelativeTime": [Function],
        "formatTime": [Function],
        "formatTimeToParts": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getDisplayNames": [Function],
          "getListFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralRules": [Function],
          "getRelativeTimeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "default.cancel": "Cancel",
          "default.delete": "Delete",
          "default.remove": "Remove",
          "default.save": "Save",
        },
        "onError": [Function],
        "textComponent": Symbol(react.fragment),
        "timeZone": undefined,
      }
    }
  >
    <TextContent>
      <div
        className="pf-c-content"
      >
        <WithLoader
          loading={true}
        >
          <Skeleton
            isDark={false}
            size="lg"
          >
            <div
              className="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </Skeleton>
        </WithLoader>
        <WithLoader
          loading={true}
        >
          <Skeleton
            isDark={false}
            size="lg"
          >
            <div
              className="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </Skeleton>
        </WithLoader>
      </div>
    </TextContent>
  </CvssVector>
</injectIntl(CvssVector)>
`;

exports[`CvssVector Should render with CVSSv3 1`] = `
<injectIntl(CvssVector)
  cvss3_metrics="CVSS:3.0/AV:A/AC:L/PR:L/UI:R/S:U/C:L/I:N/A:N"
>
  <CvssVector
    cvss3_metrics="CVSS:3.0/AV:A/AC:L/PR:L/UI:R/S:U/C:L/I:N/A:N"
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatDateToParts": [Function],
        "formatDisplayName": [Function],
        "formatHTMLMessage": [Function],
        "formatList": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatNumberToParts": [Function],
        "formatPlural": [Function],
        "formatRelativeTime": [Function],
        "formatTime": [Function],
        "formatTimeToParts": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getDisplayNames": [Function],
          "getListFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralRules": [Function],
          "getRelativeTimeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "default.cancel": "Cancel",
          "default.delete": "Delete",
          "default.remove": "Remove",
          "default.save": "Save",
        },
        "onError": [Function],
        "textComponent": Symbol(react.fragment),
        "timeZone": undefined,
      }
    }
  >
    <TextContent>
      <div
        className="pf-c-content"
      >
        <WithLoader
          loading={true}
        >
          <Skeleton
            isDark={false}
            size="lg"
          >
            <div
              className="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </Skeleton>
        </WithLoader>
        <WithLoader
          loading={true}
        >
          <Skeleton
            isDark={false}
            size="lg"
          >
            <div
              className="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </Skeleton>
        </WithLoader>
      </div>
    </TextContent>
  </CvssVector>
</injectIntl(CvssVector)>
`;

exports[`CvssVector Should render without parameters 1`] = `
<injectIntl(CvssVector)>
  <CvssVector
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatDateToParts": [Function],
        "formatDisplayName": [Function],
        "formatHTMLMessage": [Function],
        "formatList": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatNumberToParts": [Function],
        "formatPlural": [Function],
        "formatRelativeTime": [Function],
        "formatTime": [Function],
        "formatTimeToParts": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getDisplayNames": [Function],
          "getListFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralRules": [Function],
          "getRelativeTimeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "default.cancel": "Cancel",
          "default.delete": "Delete",
          "default.remove": "Remove",
          "default.save": "Save",
        },
        "onError": [Function],
        "textComponent": Symbol(react.fragment),
        "timeZone": undefined,
      }
    }
  >
    <TextContent>
      <div
        className="pf-c-content"
      >
        <WithLoader
          loading={true}
        >
          <Skeleton
            isDark={false}
            size="lg"
          >
            <div
              className="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </Skeleton>
        </WithLoader>
        <WithLoader
          loading={true}
        >
          <Skeleton
            isDark={false}
            size="lg"
          >
            <div
              className="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </Skeleton>
        </WithLoader>
      </div>
    </TextContent>
  </CvssVector>
</injectIntl(CvssVector)>
`;
